go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -0,0 +1,957 @@
+	nadlisters "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/listers/k8s.cni.cncf.io/v1"
+	frrtypes "github.com/metallb/frr-k8s/api/v1beta1"
+	frrclientset "github.com/metallb/frr-k8s/pkg/client/clientset/versioned"
+	frrlisters "github.com/metallb/frr-k8s/pkg/client/listers/api/v1beta1"
+	core "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	metaapply "k8s.io/client-go/applyconfigurations/meta/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	controllerutil "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/controller"
+	eiptypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1"
+	egressiplisters "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/apis/listers/egressip/v1"
+	ratypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/routeadvertisements/v1"
+	raapply "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/routeadvertisements/v1/apis/applyconfiguration/routeadvertisements/v1"
+	raclientset "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/routeadvertisements/v1/apis/clientset/versioned"
+	ralisters "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/routeadvertisements/v1/apis/listers/routeadvertisements/v1"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+const (
+	generateName = "ovnk-generated-"
+	fieldManager = "clustermanager-routeadvertisements-controller"
+)
+
+var (
+	errConfig = errors.New("configuration error")
+)
+
+// Controller reconciles RouteAdvertisements
+type Controller struct {
+	eipLister  egressiplisters.EgressIPLister
+	frrLister  frrlisters.FRRConfigurationLister
+	nadLister  nadlisters.NetworkAttachmentDefinitionLister
+	nodeLister corelisters.NodeLister
+	raLister   ralisters.RouteAdvertisementsLister
+
+	frrClient frrclientset.Interface
+	nadClient nadclientset.Interface
+	raClient  raclientset.Interface
+
+	eipController  controllerutil.Controller
+	frrController  controllerutil.Controller
+	nadController  controllerutil.Controller
+	nodeController controllerutil.Controller
+	raController   controllerutil.Controller
+}
+
+// NewController builds a controller that reconciles RouteAdvertisements
+func NewController(wf *factory.WatchFactory, ovnClient *util.OVNClusterManagerClientset) *Controller {
+	c := &Controller{
+		eipLister:  wf.EgressIPInformer().Lister(),
+		frrLister:  wf.FRRConfigurationsInformer().Lister(),
+		nadLister:  wf.NADInformer().Lister(),
```

Comment:
We shall only initialize `nadLister` if `multi-net` is enabled. Otherwise, it will get `nil pointer` from `NADInformer()`.

Reply:
This controller itself is only initialized if multi-net is enabled:
```
func IsRouteAdvertisementsEnabled() bool {
	// for now, we require multi-network to be enabled because we rely on NADs,
	// even for the default network
	return config.OVNKubernetesFeature.EnableMultiNetwork && config.OVNKubernetesFeature.EnableRouteAdvertisements
}
```
I think @pliurh comment was done before I made that decision. I think this comment actually triggered me to make the decision. So already addressed!